### PR TITLE
Add ElasticSearch to Docker Compose

### PR DESCRIPTION
**Pull request by citizen428**
## What type of PR is this? (check all applicable)

- [X] Developer experience

## Description

We recently introduced ES to our code base but haven't updated our Docker workflow accordingly yet. 

## Related Tickets & Documents

Discussed in Slack with @mstruve 

## Added to documentation?

- [X] docs.dev.to

NOTE: this needs #5976 to be merged first in order to work correctly. I decided to split it into 2 PRs in case ES needs more work, so we can at least merge the fix first.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -69,3 +69,6 @@ docs/.static/api/
 
 # asdf
 .tool-versions
+
+# ES docker volume
+docker_data/
```
Comment (citizen428) [inline .gitignore]: Added this to ignore the shared volume for ES.

File: docker-compose.yml [hunk 2]
```diff
@@ -33,3 +35,23 @@ services:
     image: "redis"
     ports:
       - "6379:6379"
+  elasticsearch:
+    image: docker.elastic.co/elasticsearch/elasticsearch:7.5.2
+    container_name: elasticsearch
+    environment:
```
Comment (citizen428) [inline docker-compose.yml]: These settings came from quite a bit of trial and error and reading (often outdated) blog posts, so let's not change them too much as long as they work.

File: docker-compose.yml [hunk 3]
```diff
@@ -33,3 +35,23 @@ services:
...
+      - cluster.name=docker-cluster
+      - bootstrap.memory_lock=true
+      - "ES_JAVA_OPTS=-Xms512m -Xmx512m"
+      - "discovery.type=single-node"
+      - xpack.security.enabled=false
```
Comment (citizen428) [inline docker-compose.yml]: @mstruve Just turned off some things that I thought we wouldn't need in local development.

File: docs/installation/docker.md [hunk 4]
```diff
@@ -84,3 +85,11 @@ Please execute the script itself to view all additional options:
 ```shell
 ./docker-run.sh
 ```
+
+## Known Problems & Solutions
+
+- Should you experience problems with the ElasticSearch container, try to
```
Comment (citizen428) [inline docs/installation/docker.md]: Figured it won't hurt to include this information.

**Comment by citizen428** [comment]
@mstruve I know you mentioned that Docker isn't your forte, but maybe you can at least check the ES settings etc. Here's some proof of work:

*Search setup*

```sh
$ docker-compose run web rails search:setup
Starting devto_db_1 ... done
Starting devto_redis_1 ... done
2020-02-11 03:43:29 +0000: HEAD http://elasticsearch:9200/tags_development [status:404, request:0.387s, query:N/A]
2020-02-11 03:43:29 +0000: <
2020-02-11 03:43:29 +0000: [404]
2020-02-11 03:43:32 +0000: PUT http://elasticsearch:9200/tags_development [status:200, request:2.789s, query:n/a]
2020-02-11 03:43:32 +0000: > {"settings":{"index":{"number_of_shards":1,"number_of_replicas":0}}}
2020-02-11 03:43:32 +0000: < {"acknowledged":true,"shards_acknowledged":true,"index":"tags_development"}
2020-02-11 03:43:32 +0000: PUT http://elasticsearch:9200/tags_development/_alias/tags_development_alias [status:200, request:0.350s, query:n/a]
2020-02-11 03:43:32 +0000: < {"acknowledged":true}
2020-02-11 03:43:33 +0000: PUT http://elasticsearch:9200/tags_development_alias/_mapping [status:200, request:0.363s, query:n/a]
2020-02-11 03:43:33 +0000: > {"dynamic":"strict","properties":{"id":{"type":"keyword"},"name":{"type":"text","fields":{"raw":{"type":"keyword"}}},"hotness_score":{"type":"integer"},"supported":{"type":"boolean"},"short_summary":{"type":"text"}}}
2020-02-11 03:43:33 +0000: < {"acknowledged":true}
```

*Connecting from web container to ES container*

```sh
$ docker-compose run web sh
Starting devto_redis_1 ... done
Starting elasticsearch ... done
Starting devto_db_1    ... done
/usr/src/app # apk add curl
fetch http://dl-cdn.alpinelinux.org/alpine/v3.10/main/x86_64/APKINDEX.tar.gz
fetch http://dl-cdn.alpinelinux.org/alpine/v3.10/community/x86_64/APKINDEX.tar.gz
(1/1) Installing curl (7.66.0-r0)
Executing busybox-1.30.1-r3.trigger
OK: 376 MiB in 134 packages
/usr/src/app # echo $ELASTICSEARCH_URL
http://elasticsearch:9200
/usr/src/app # curl http://elasticsearch:9200
{
  "name" : "3218d5317ecb",
  "cluster_name" : "docker-cluster",
  "cluster_uuid" : "M-71G4UFREWgxpbXG9A8ag",
  "version" : {
    "number" : "7.5.2",
    "build_flavor" : "default",
    "build_type" : "docker",
    "build_hash" : "8bec50e1e0ad29dad5653712cf3bb580cd1afcdf",
    "build_date" : "2020-01-15T12:11:52.313576Z",
    "build_snapshot" : false,
    "lucene_version" : "8.3.0",
    "minimum_wire_compatibility_version" : "6.8.0",
    "minimum_index_compatibility_version" : "6.0.0-beta1"
  },
  "tagline" : "You Know, for Search"
}
```

File: docker-compose.yml (hunk 2 above)
Comment (mstruve) [reply docker-compose.yml]: Don't try to fix something that ain't broke 😉

File: docker-compose.yml [hunk 5]
```diff
@@ -33,3 +35,23 @@ services:
...
     ports:
       - "6379:6379"
+  elasticsearch:
+    image: docker.elastic.co/elasticsearch/elasticsearch:7.5.2
+    container_name: elasticsearch
+    environment:
+      - cluster.name=docker-cluster
+      - bootstrap.memory_lock=true
+      - "ES_JAVA_OPTS=-Xms512m -Xmx512m"
+      - "discovery.type=single-node"
```
Comment (mstruve) [inline docker-compose.yml]: May want to change this in the future to allow for multiple nodes but this is good for now.

File: docs/installation/docker.md (hunk 4 above)
Comment (mstruve) [reply docs/installation/docker.md]: Just a random PSA, its Elasticsearch, no capital on the search :)

File: docker-compose.yml (hunk 5 above)
Comment (citizen428) [reply docker-compose.yml]: Yes, I thought about that, but figured let's keep it simple for now, we shouldn't need bigger clusters for the dev workflow.

**Pull request merged**